adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -817,32 +818,36 @@ void handlePublishedMessage(final MqttContext context) {
         }
 
         final MetricsTags.QoS qos = MetricsTags.QoS.from(ctx.message().qosLevel().value());
-        return uploadMessage(
+        final Future<TenantObject> tenantTracker = getTenantConfiguration(tenant, ctx.getTracingContext());
+
+        return tenantTracker.compose(tenantObject -> uploadMessage(
                 Objects.requireNonNull(ctx),
-                Objects.requireNonNull(tenant),
+                tenantObject,
                 Objects.requireNonNull(deviceId),
                 Objects.requireNonNull(payload),
                 getTelemetrySender(tenant),
-                ctx.endpoint()
-        ).map(success -> {
-            metrics.reportTelemetry(
-                    ctx.endpoint(),
-                    ctx.tenant(),
-                    MetricsTags.ProcessingOutcome.FORWARDED,
-                    qos,
-                    payload.length(),
-                    ctx.getTimer());
-            return success;
-        }).recover(t -> {
-            metrics.reportTelemetry(
-                    ctx.endpoint(),
-                    ctx.tenant(),
-                    ProcessingOutcome.from(t),
-                    qos,
-                    payload.length(),
-                    ctx.getTimer());
-            return Future.failedFuture(t);
-        });
+                ctx.endpoint()))
```

Comment:
does it make sense to postpone the null checks to *after* the Future for getting the tenant configuration has completed? IMHO the null checks should be moved to the beginning of the method ...

Reply:
I have moved it to the beginning.

Reply:
have you?

Reply:
I will push the changes along with the one for the above comment.

Reply:
@sophokles73 I have pushed the changes. Pls have a look.